module/VuFind/src/VuFind/Controller/Feature/OnlinePaymentTrait.php

```diff
@@ -0,0 +1,329 @@
+        $view->lastPayment = $lastPayment;
+
+        $paymentInProgress = $paymentService->getPaidPaymentInProgressForPatron($patron['cat_username']);
+        if (
+            $pay
+            && $paymentDetails['payable']
+            && $paymentDetails['amount']
+            && !$paymentInProgress
+        ) {
+            // Check CSRF:
+            $csrfValidator = $this->serviceLocator->get(\VuFind\Validator\CsrfInterface::class);
+            $csrf = $this->getRequest()->getPost()->get('csrf');
+            if (!$csrfValidator->isValid($csrf)) {
+                $this->flashMessenger()->addErrorMessage('Payment::error_payment_request_failed');
+                header('Location: ' . $this->getServerUrl('myresearch-fines'));
+                exit();
+            }
+            // After successful token verification, clear list to shrink session and
+            // ensure that the form is not re-sent:
+            $csrfValidator->trimTokenList(0);
+
+            // Payment requested, do preliminary checks:
+            if ($paymentInProgress) {
+                $this->flashMessenger()->addErrorMessage('Payment::error_payment_request_failed');
+                header('Location: ' . $this->getServerUrl('myresearch-fines'));
+                exit();
+            }
+            if (
+                (($paymentConfig['exactBalanceRequired'] ?? true)
+                || !empty($paymentConfig['creditUnsupported']))
+                && !$selectFees
+                && $onlinePaymentManager->getStoredPayableAmount($patron) !== $paymentDetails['amount']
+            ) {
+                // Fines updated, redirect and show updated list.
+                $this->flashMessenger()->addErrorMessage('Payment::error_fines_changed');
+                header('Location: ' . $this->getServerUrl('myresearch-fines'));
+                exit();
+            }
+            $returnUrl = $this->getServerUrl('myresearch-fines');
+            // Include language in notify url because it's a back-channel request that
+            // doesn't have access to user's session:
+            $notifyUrl = $this->getServerUrl('home') . 'AJAX/onlinePaymentNotify?lng='
+                . urlencode($this->getTranslatorLocale());
+
+            // Start payment
+            try {
+                $onlinePaymentManager->startPayment(
+                    $returnUrl,
+                    $notifyUrl,
+                    $user,
+                    $patron,
+                    $paymentDetails['amount'],
+                    $paymentDetails['fines'],
+                    'local_payment_id'
+                );
+            } catch (PaymentException $e) {
+                $this->flashMessenger()->addErrorMessage($e->getMessage());
+            }
+            // We should only end up here on error, but redirect always just in case
+            // the payment handler somehow misbehaves:
+            header('Location: ' . $this->getServerUrl('myresearch-fines'));
+            exit();
```

Comment:
If this function almost always exits, it does feel to me like returning something the calling code could use might feel cleaner...